keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -0,0 +1,996 @@
+3. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by kube-scheduler.
+4. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by different
+   schedulers.
+5. Avoid premature preemptions of already running pods in case a higher priority gang will be rejected.
+6. Support gang-level (or workload-level in general) level preemption (if pods form a gang also
+   from a runtime perspective, they can't be preempted individually).
+7. Updating workload status and triggering rescheduling when a gang failed binding in the all-or-nothing
+   fashion.
+8. Support gang-scheduling even if part of the infrastructure needs to be provisioned (by Cluster
+   Autoscaler, Karpenter or other solutions).
+
+Addressing all these requirements in a single shot would be a huge change, so as part ot this KEP we
+will only focus on a subset of those. However, we very briefly sketch the path towards the vision to
+ensure that this KEP is moving in the right direction.
+
+For `Alpha`, we are focusing on introducing the concept of the `Workload` and plumbing it into
+kube-scheduler in the simplest possible way. We will implement a new plugin implementing the following
+hooks:
+- PreEnqueue - used as a barrier to wait for the `Workload` object and all the necessary pods to be
+  observed by the scheduler before even considering them for actual scheduling
+- WaitOnPermit - used as a barrier to wait for the pods to be assigned to the nodes before initiating
+  potential preemptions and their bindings
+
+This seems to be the simplest possible implementation to address the requirement (1). We are consciously
+ignoring the rest of the requirements for `Alpha` phase.
+
+
+For `Beta`, we want to also touch requirements (2) and (3) by extending the scheduling framework with
+a new dedicated phase (tentatively called Workload). In that phase,
+kube-scheduler will be looking at all pods from a gang (part of `Workload`) and compute the placement
+for all of these pods in a single scheduling cycle. Those placements will be stored only in-memory and
+block the required resources from scheduling. Tentively we plan to use `NominatedNodeName` field for it.
+After that, pods will go through regular pod-by-pod scheduling phases (including Filter and Score)
+with a nomination as a form of validation the proposed placement and execution of this placement decision.
+Therefore we expect the order of processing pods won't ever be important, but all-or-nothing nature of
+gangs will be preserved while advancing through the further steps of the binding process.
+
+While we will not target addressing "optimal enough" part of requirement (2), we will assure that we
+can process all gang pods together. The single scheduling cycle and blocking resources in beta
+will address the requirement (3).
+
+We will also introduce delayed preemption by moving it after `WaitOnPermit` phase. Together with
```

Comment:
What if a pod deletion takes long? The current scheduler mitigates it by just letting a preemptor pod go anywhere else (if there's an empty space, probably made by some other pod termination), while a deletion isn't completed.